### ISSUE TITLE
The purpose of the :column option and its possible enhancement (replacement)

### ISSUE THREAD

**Issue by skalee** [riboseinc/attr_masker#22]
This question is somewhat a continuation of my discussion with @ronaldtse in the #5, which consisted of following two comments.

>*skalee*: I wonder what the `:column_name` option is for. Apparently it overrides the column name when it's different from the attribute name. However, I cannot think of a good reason to support it at the moment. It isn't present in the original gem, and has been added in commit 82cab53, so I bet it is used by one of your projects. Therefore, I ask you for some concrete example or green light on its removal.

> *ronaldtse*: @skalee I suspect it was used to support attributes that uses different column names like attr_encrypted does, but it might no longer be used. @ribose-jeffreylau should be able to clarify this with authority since he wrote it.

However, after some thinking I am coming to conclusion that `:column_name` option is too limited to be useful. In most cases, the attribute name is the same as the column name. I can imagine rare cases when this assumption is false, like attachment attribute handled by Paperclip or Carrierwave gem, or some serialized data stored in a column which name is suffixed. However in some cases things tend to go even more complex, for example Paperclip [may use up to 4 columns](https://github.com/thoughtbot/paperclip#usage). Passing single column name to that parameter is not enough.

I suggest renaming `:column_name` to `:column_names` and allow passing an array of column names. For example,

```ruby
attr_masker :photo, column_names: %i[photo_file_name photo_content_type]
```

would update two columns.

I am going to implement it this way unless someone opposes.

**Comment by skalee** [comment]
@ribose-jeffreylau Your input here is very welcome. I'd really like to know the usage example for the `column_name` option before I make any radical changes.

**Comment by skalee** [comment]
Okay, so I thought about it a bit, and I got two ideas:

First of all, how it works now:

1. The masker return value is simply passed to the SQL `UPDATE` command.  Model attribute writers are not involved.
2. By default, name of the column to be updated is derived from the attribute name, but it can be overwritten with `column_name` option.

We can easily enhance it in a following way:

1. An array of column names can be passed as `column_names` option.  If so, the masker's return value is expected to be an array as well, and its subsequent elements are update values for corresponding columns.

Example:

Here we're setting some agreed password for every user so that developers can log in as anyone they want.  Password digest and salt are stored in separate columns.

```ruby
class User
  attr_masker :password, column_names: ["password_digest", "password_salt"], masker: proc do
    new_password = "verysecret"
    salt = generate_random_salt()
    digest = encrypt_password(new_password, salt)
    [digest, salt]
  end
end
```

The masker's return value will be used in update operation like this one:

```ruby
User.update(user.id, {password_digest: masker_retval[0], password_hash: masker_retval[1]})
```

However I am also thinking about a more radical change:

1. The masker's return value is assigned to model's attribute via respective attribute writer method.
2. The attribute writer may have arbitrary side effects.  In particular, it may alter more than one database fields (see example below).
3. However only changes made to fields specified in `column_names` option are saved.  All other changes are discarded.  The `column_names` option defaults to one-element array containing the attribute name.

Example:

```ruby
class User
  attr_masker :password, masker: proc { "verysecret" }, column_names: ["password_hash", "password_salt"]

  def password=(new_password)
    self.password_salt = salt = generate_random_salt()
    self.password_hash = encrypt_password(new_password, salt)
  end
end
```

Consequences of this approach are as follows:
1. It must be user's responsibility to ensure that the attribute writer is safe to use.  For example, it could perform a database update (and that will happen before other attributes get reverted!), or some junk could be uploaded to S3, overwriting production pictures.  On the other hand, `:if`, and `:unless` filters may also call model methods, so maybe it's not that bad.  (But typically they will rely on getter methods and are far less likely to break anything).
2. Since we're working on models, it should be possible to validate them before saving.

**Comment by ribose-jeffreylau** [comment]
@skalee 
For our current use case, it is actually very similar to the one you just posted above.
We use attr_encrypted alongside attr_masker.  Using attr_encrypted means there would be some columns that are renamed to `encrypted_old_column_name` instead of just `old_column_name`.
In our test environment, we also want to be able to mask the column `old_column_name`.  This is where `:column_name` comes into play.
Your proposed solution looks like it could handle this situation well.

**Comment by ronaldtse** [comment]
Thanks @ribose-jeffreylau for the clarification, @skalee let's proceed, thanks!

**Comment by ronaldtse** [comment]
@skalee I agree that the solution to #51 is this. The attribute writer can support "super" so that all necessary attribute writer modifiers are applied prior to masking. 

I thought the tricky part is how to ensure attr_masker is in the attribute-writers chain, but that said all overrides to the attribute writer would technically have to call super so it won't be an issue. 

Let's go ahead, thanks!

**Comment by skalee** [comment]
As for now, attr_masker does not use attribute writers at all.

**Comment by ronaldtse** [comment]
Got it, let's do it. 👍

**Comment by skalee** [comment]
The `column_name` option has been introduced in #53.
